src/dali/plugin/input/rest/kucoin_futures.py

```diff
@@ -0,0 +1,288 @@
+            try:
+                ledger_transactions = self._client.futuresPrivateGetTransactionHistory({_START_AT: start_at, _END_AT: start_at + _TS_INCREMENT})
+                data_list = ledger_transactions[_DATA][_DATA_LIST]
+                has_more = ledger_transactions[_DATA][_HAS_MORE]
+
+                for item in data_list:
+                    items.append(item)
+                    offset = item[_OFFSET]
+
+                # https://github.com/ccxt/ccxt/issues/10273
+                # enableRateLimit is not working for kucoin
+                time.sleep(0.15)
+                retries = 0
+
+                while has_more and (retries < 4):
+                    try:
+                        ledger_transactions = self._client.futuresPrivateGetTransactionHistory(
+                            {_START_AT: start_at, _END_AT: start_at + _TS_INCREMENT, _OFFSET: offset}
+                        )
+                        data_list = ledger_transactions[_DATA][_DATA_LIST]
+                        has_more = ledger_transactions[_DATA][_HAS_MORE]
+
+                        for item in data_list:
+                            items.append(item)
+                            offset = item[_OFFSET]
+
+                        time.sleep(0.15)
+
+                    except RateLimitExceeded:
+                        self._client.sleep(13000)
+                        retries += 1
+
+                if retries >= 4:
+                    raise Exception("Failed to fetch ledger transactions after 4 retries")
+
+                start_at += _TS_INCREMENT
+                offset = 1
+                has_more = False
+                retries = 0
+
+            except RateLimitExceeded:
+                self._client.sleep(13000)
+                retries += 1
+
+            except Exception as exception:
+                raise exception
+
+        if retries >= 4:
+            raise Exception("Failed to fetch ledger transactions after 4 retries")
+
+        return items
+
+    def _process_data(
+        self,
+        items: List[Dict[str, Any]],
+        in_transactions: List[InTransaction],
+        out_transactions: List[OutTransaction],
+        # intra_transactions: List[IntraTransaction],
+    ) -> None:
+        for item in items:
+            timestamp_value: int = item[_TIME]
+            timestamp: str = datetime.fromtimestamp(float(timestamp_value) / 1000.0, tz.tzutc()).strftime("%Y-%m-%d %H:%M:%S.%f%z")
+            transaction_type: str = item[_TYPE]
+            amount: str = str(item[_AMOUNT])
+            currency: str = item[_CURRENCY]
+            offset: int = item[_OFFSET]
+
+            # internal tranfer between kucoin and kucoin futures. Ignore
+            if transaction_type in {_TRANSFER_IN, _TRANSFER_OUT}:
+                continue
+
+            # To do: Currently Kucoin futures allows BTC and USDT deposits / withdrawals through blockchain wallets
+            if transaction_type in {_DEPOSIT, _WITHDRAWAL}:
```

Comment:
It looks like kucoin_futures on CCXT supports `fetchDeposits()` and `fetchWithdrawals()`. If you have time, you can implement the `pagination_detail_set`s for those and the abstract plugin should do the rest for you.

Reply:
Yes, let's use the full functionality of AbstractCcxtInputPlugin. This will enable two things:
* less duplicated code;
* find and fix more bugs in AbstractCcxtInputPlugin.

Reply:
Left a comment to implement fetchDeposit and fetchWithdrawals as to do in the future. I've decided not to implement it at the moment as I don't currently have deposit/withdrawal transactions and won't be able to test it.